### DISCUSSION TITLE
Calling multiple JS functions consecutively from CPP, does using official encapsulation methods lead to performance waste?

### DISCUSSION THREAD

**Discussion by sxstd001** [emscripten-core/emscripten#21871] [Q&A]
**Is the performance of A superior to that of B?**
Is that:  A involves only one context switch, whereas B requires multiple consecutive context switches?

A:
```cpp
	EM_ASM({
		gl.useProgram($0);  
		gl.uniformMatrix3fv($1, false, $2)
                gl.xxxx...
	}, program, u_mvp, matrix);
```
B:
```cpp
  emscripten::glUseProgram(program);
  emscripten::glUniformMatrix3fv(u_mvp, false, matrix);
  emscripten::gl_xxxx....
```

**Comment by sbc100** [comment]
You are asking here about the code of calling from Wasm to JavaScript.  This tends to vary a little between browsers, but should be very fast.  Especially when you are making calls into WebGL or WebGPU like this I imagine the cost of the graphics API call makes the cost of the Wasm/JS crossing insignificant.   Obviously the only what to really know this is to measure of the browser(s) you are targetting.  In general I would advise against this type of optimization since its likely to be premature, and also change over time as browsers improve.    @kainino0x WDYT?
